src/vision/status_quo/barbara_tries_unix_socket.md

```diff
@@ -0,0 +1,314 @@
+        .unwrap();
+    Ok(res)
+}
+
+```
+
+## Nice simple query with high-level reqwest
+
+Barbara do HTTP GET request using TCP socket with reqwest and it works fine, everything is easy.
+
+```rust
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let res = reqwest::get("http://127.0.0.1:3000").await?;
+    println!("{}", res.text().await?);
+    Ok(()) 
+}
+```
+
+## Unix sockets for performance
+
+One day, Barbara heard that using unix socket can provide a better performance by using unix socket when both the server and client is on the same machine, so Barbara decided to try it out.
+
+Barbara starts porting the server code to use unix socket, it was a no brainer for Barbara at least. Barbara changed `TcpListener::bind("127.0.0.1:3000").await?` to `UnixListener::bind("/tmp/socket")?` and it works like a charm.
+
+Barbara search through reqwest doc and github issues to see how to use unix socket for reqwest. Barbara found https://github.com/seanmonstar/reqwest/issues/39#issuecomment-778716774 saying reqwest does not support unix socket but hyper does with an example, which is a lower-level library. Since reqwest is so easy and porting hyper server to use unix socket is easy, Barbara thinks low-level hyper library should be easy too.
+
+## The screen stares at Barbara
+
+Barbara wrote some code according to the comments Barbara saw and read some docs like what is `handshake` to roughly know what it does. Barbara compile and it shows a warning, the `connection` variable is not used:
+```
+warning: unused variable: `connection`
+ --> src/main.rs:9:30
+  |
+9 |     let (mut request_sender, connection) = conn::handshake(stream).await?;
+  |                              ^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_connection`
+  |
+  = note: `#[warn(unused_variables)]` on by default
+```
+
+Barbara then runs the program. Barbara was stares at the screen and the screen stares at her. Barbara waited and ... it was stuck. So Barbara decides to look at the logs and run it again with `env RUST_LOG=trace cargo r`, and it was indeed stuck, but not sure where.
+```
+ TRACE mio::poll > registering event source with poller: token=Token(0), interests=READABLE | WRITABLE
+```
+
+Barbara try adding `println!` all over the code but it was still stuck, so Barbara try asking for help. Thanks to the welcoming Rust community, Barbara got help quickly in this case. It seemed like Barbara missed the `connection` which is a culprit and it was in the parent module of the docs Barbara read.
+
+Barbara added the missing piece to `.await` for the `connection`, all the while Barbara thought it will work if it was `.await`-ed but in this case having required to await something else to work is a surprise.
+
+```rust
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("error: {}", e);
+        }
+    })
+    
+    let request = ...
+```
```

Comment:
Spawning a new task looks a little strange at first, but this is indeed what the hyper docs recommend.  Is that where you/Barbara got the idea to await the connection in this way?  If so, it would probably be good to mention that.

Reply:
I didn't know it is required to do it this way, but it does seemed weird. I didn't found the relevant part in docs until someone point me to it which is in the parent module. I got the idea until someone told me that it is required.